python/GafferSceneUI/HierarchyView.py

```diff
@@ -356,3 +359,121 @@ def _updateFromPathFilter( self ) :
 	def __patternEditingFinished( self, widget ) :
 
 		self.pathFilter().setMatchPattern( self.__patternWidget.getText() )
+
+##########################################################################
+# _ExpansionWidget
+##########################################################################
+
+class _ExpansionWidget( GafferUI.Widget ) :
+
+	def __init__( self ) :
+
+		button = GafferUI.MenuButton(
+			image = "locationExpanded.png",
+			hasFrame = False,
+			toolTip = "Expansion",
+			menu = GafferUI.Menu( Gaffer.WeakMethod( self.__menuDefinition ), title = "Expansion" )
+		)
+
+		GafferUI.Widget.__init__( self, button )
+
+	def __menuDefinition( self ) :
+
+		menu = IECore.MenuDefinition()
+
+		menu.append(
+			"/Bookmarks",
+			{
+				"subMenu" : Gaffer.WeakMethod( self.__bookmarkMenuDefinition )
+			}
+		)
+
+		return menu
+
+	def __bookmarkMenuDefinition( self ) :
+
+		scriptNode = self.ancestor( GafferUI.ScriptWindow ).scriptNode()
+		readOnly = Gaffer.MetadataAlgo.readOnly( scriptNode )
+
+		menuDefinition = IECore.MenuDefinition()
+
+		bookmarks = sorted( GafferSceneUI.ScriptNodeAlgo.visibleSetBookmarks( scriptNode ) )
+		if bookmarks :
+			for b in bookmarks :
+				menuDefinition.append(
+					"/" + b,
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__restore ), b ),
+						"checkBox" : GafferSceneUI.ScriptNodeAlgo.getVisibleSetBookmark( scriptNode, b ) == GafferSceneUI.ScriptNodeAlgo.getVisibleSet( scriptNode ),
+					}
+				)
+			menuDefinition.append( "/RestoreBookmarkDivider", { "divider" : True } )
+
+			for b in bookmarks :
+				menuDefinition.append(
+					"/Save As/" + b,
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__saveAs ), b ),
+						"active" : not readOnly,
+					}
+				)
+				menuDefinition.append(
+					"/Delete/" + b,
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__delete ), b ),
+						"active" : not readOnly,
+					}
+				)
+			menuDefinition.append( "/Save As/Divider", { "divider" : True } )
+		else :
+			menuDefinition.append( "/No Bookmarks Available", { "active" : False } )
+			menuDefinition.append( "/NoBookmarksDivider", { "divider" : True } )
+
+		menuDefinition.append(
+			"/Save As/New Bookmark...",
+			{
+				"command" : functools.partial( Gaffer.WeakMethod( self.__save ) ),
+				"active" : not readOnly,
+			}
+		)
+
+		return menuDefinition
+
+	def __save( self, *unused ) :
+
+		scriptWindow = self.ancestor( GafferUI.ScriptWindow )
```

Comment:
Parenting to the ScriptWindow is a little odd if the HierarchyView is in a different window. Perhaps just `self.ancestor( GafferUI.Window )`?

Reply:
Changed in c7602bd.